### ISSUE TITLE
Using factory attrs in route handlers

### ISSUE THREAD

**Issue by lmcardle** [samselikoff/ember-cli-mirage#231]
When I am creating a new record, from a POST request, the record created does not contain all properties of the factory that is defined.  For example.

I have a page for new user, the post request sends only the user password and username.  However, the factory has a few additional properties:
- username
- password
- createdAt
- updatedAt

As I said, the request.requestBody only contains "username" and "password" and when I create a new record, the createdAt and updatedAt properties are not created

```
var body = { username: 'sam', password: '1234' }
var newUser = db.users.insert(body)
console.log(newUser) // { id: 123, username: 'sam', password: '1234' }
```

**Comment by samselikoff** [comment]
The factory definition is only used to "seed" your db with initial values, either in development `/scenarios/default.js` or within a test. `request.requestBody` contains the attrs sent over from your Ember app, so if your form only has those two, that's all that should be there.

There's nothing like a default `createdAt` and `updatedAt` on db records (right now), so you'd have to set those yourself in the route handler.

**Comment by lmcardle** [comment]
Thanks. That's unfortunate. I have a lot more properties than created at
and updated at. There are often properties that have a default in the
database if the user does not supply a value or another case would be a
property in the database that is calculated for the user.

Is there any way to get access to the factory to loop over properties and
create a new value?

**Comment by samselikoff** [comment]
This is not really the purpose of factories so I would discourage heading down this path. Same reason we don't want to use factory_girl factories in rails controllers to create data.

There will be a model introduced shortly, perhaps a constructor method there would satisfy the requirement? Or if this is all at the db level, we could have something like default values...though it's all just a simple Object storage, no schema or anything

**Comment by lmcardle** [comment]
Do you have any info on the model solution you are thinking of? I can take a look and let you know if that would work.  I'm for as simple solution as possible.  

Anything I can do to help with on moving this forward.

**Comment by samselikoff** [comment]
A model layer/simple orm is actually already merged into Mirage, paving the way for serializers and factory relationships. For those features, it will be a simple declaration like

```js
// mirage/models/user.js
export default Mirage.Model.extend({
  posts: Mirage.hasMany()
});
```

We could add a constructor if you wanted to set some defaults:

```js
// mirage/models/user.js
export default Mirage.Model.extend({
  constructor() {
    this.createdAt = new Date().toString();
  },
  posts: Mirage.hasMany()
});
```

or something like this. I'm not sure, fyi this is not super high priority for me right now so you may just want to make a function or object that you can share with your routes and your factories for now.

**Comment by lmcardle** [comment]
I have gotten around this issue for now because it looks like "server" is a global.

```
// Data Passed to the POST Request
{
  data: {
    name: 'liam'
  }
}
```

```
// Data Needed to be returned by the server
{
  data: {
    name: 'Liam',
    role: 'general'   // note this is the default role given if non specified
  }
}
```

```
// config.js
this.post('/users', function(db, request) {
  let body = JSON.parse(request.requestBody);
  let newUser = server.create('user', body.data);
  return { data: newUser };
});
```

While this approach works, I've got a few questions:
* Is there any risk in using "server.create" here?
* Why is "server" a global?  I was initially looking for a way to grab the server instance but stumbled upon it being a global?

**Comment by lmcardle** [comment]
Sorry...my second question is irrelevant, server is not a global rather it is "this". Just continuing to dig through the code.  :)

**Comment by samselikoff** [comment]
Yep exactly `server` is `this`. `server.create` is not intended to be used within a route handler but, I'm backed up for now so if you can hack it, go for it :) I think the solution here is to have a `constructor()` method on the model.

**Comment by pzuraq** [comment]
@samselikoff We would actually like to be able to have this type of functionality for POST-ing. Some our systems only take a single field and then populate the rest of the data from other sources on the server side. It would be convenient to use Factories and faker.js to be able to achieve a similar effect in our mocks/tests.

**Comment by samselikoff** [comment]
@pzuraq hm. you can use faker anywhere you like, just import it. So you could use it in your route handler, to simulate getting data from another source.

I'll have to think more about sharing though. Factories are pretty explicitly for creating test data, and you wouldn't use factory_girl in your rails app to seed a model with default data.

**Comment by pzuraq** [comment]
Good point re: importing faker.js and assigning properties directly. I think that's a better overall solution, this is definitely not the usual use case for us, normally we create and populate our models in the POST itself.

**Comment by samselikoff** [comment]
Closing due to inactivity.

**Comment by john-griffin** [comment]
A constructor like function on the model would be super useful for setting default values such as created at. The model feels like a better place for this rather than injecting default values in at the route level, especially for apps using mirage shorthands and plain JSON API.

**Comment by samselikoff** [comment]
I agree with @john-griffin that a constructor-like function on the model would be the right place for this...

It has become a bit complicated by the fact that we introspect Ember Data models now, so I think many projects don't have a `models` directory anymore.

However you can always define Mirage models and just re-define the relationships, so if we did want to add additional APIs that are appropriate on the Model layer I think that would be ok.

We could even add an `introspectRelationships()` decorator (something like that) so folks could have Model files if they need them, but grab the relationships from their Ember Data schema.

I think the path forward here is something like that – and then if we start using ES6 classes folks can just use the built-in `constructor` to do custom init logic.

I'd be happy to help anyone who wants to work on this!

**Comment by cwagner22** [comment]
Unfortunately the `server.create('user', body.data);` approach only works for `environment: "test"`
I wish there was another alternative.

**Comment by manzaloros** [comment]
I came across this thread when I realized in my mock route handler, `schema.create('entity-that-has-a-model', {...someAttributes})` does not actually use the factory associated with that model, so the returned entity will only have the attributes I specify, rather than `{allMockAttributesFromFactory, ...someAttributes}`.

However, using `schema.create('entity-that-has-a-model', {...someAttributes})` does what I expect it to do.

This feels like a hack, but how else am I supposed to mock a `POST` route that accepts a partial object and mocks the rest of the attributes for the created entity?
